readme.txt
=== Customizable Captcha and Contact us ===
Contributors: Sreekanth Dayanand , http://www.outsource-online.net/
Donate link: http://www.outsource-online.net/osolmulticaptcha-simplest-php-captcha-for-html-forms.html
Tags: Captcha, Contact Us ,AJAX
Requires at least: 3.3
Tested up to: 6.0
Stable tag: 0.7
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

Description: Plugin to add captcha to core wordpress forms and additional option for contact us page with ajax captcha verification.

== Description ==

Description: Plugin to add captcha to core wordpress forms and additional option for contact us page.Just need to insert the code [cccontact] in a page where you want to show contact us form. 

Thanks for downloading and installing my plugin. You can show your appreciation and support future development by donating.

For more info visit: 
https://wordpress.org/plugins/customizable-captcha-and-contact-us-form/
and
http://www.outsource-online.net/osolmulticaptcha-simplest-php-captcha-for-html-forms.html

== Installation ==

1.Download the plugin
2. From WordPress 4 , Admin Panel >> Plugins, Seach for "Customizable Captcha and Contact us". Click "install" & the "activate" . For older versions, Upload plugin to the `/wp-content/plugins/` directory. 
3. Activate the plugin through the 'Plugins' menu in WordPress
4. Set the options for captcha and Contactus form in Settings Tab as in the screenshots 6 and 7
5. To use contact us form functionality,  insert the code [cccontact] in a page/post where you want to show contact us form.

== Frequently asked questions ==

= A question that someone might have =

An answer to that question.

== Screenshots ==

1. screenshot-1-login.png
2. screenshot-2-register.png
3. screenshot-3-post-comment.png
4. screenshot-4-lost-password.png
5. screenshot-5-contact-us-form.png
6. screenshot-6-admin-captcha-settings.png
7. screenshot-6-admin-contact-form-settings.png

== Changelog ==



== Upgrade notice ==



== Arbitrary section 1 ==



changeLog.md
# Change Log

## Stable Tag 0.9

### 08-10-2022

1. corrected bug in MVC/ExtraClasses/OSOLMultiCaptcha.php by appropriately setting `$this->fontPNGLocation`
```
$this->fontPNGLocation = realpath(dirname(__FILE__).$this->DS.'..'.$this->DS.'..'.$this->DS)
```
2. Fixed bug in MVC/Views/ContactusView.php adding a slash to stdClass
```
$var2Display = new \stdClass();
```
3. in custCaptchaContact.php changed 
```
	add_action( 'lostpassword_form', 'include_cust_captcha_lostpassword' );
	add_action( 'lostpassword_post', 'include_cust_captcha_lostpassword_post', 10, 3 );
```
to
```
	add_action( 'lostpassword_form', [$OSOLCCC_Frontend_inst,'include_cust_captcha_lostpassword'] );
	add_action( 'lostpassword_post', [$OSOLCCC_Frontend_inst,'include_cust_captcha_lostpassword_post'], 10, 3 );
```

4. in MVC/Views/ContactusView.php
added inside `function getCaptchaBlockHTML($forForm = "login")`
```
		$REQUEST_URI = filter_input(INPUT_SERVER, 'REQUEST_URI'); 
		//echo $REQUEST_URI . " pos is ". strpos( $REQUEST_URI, '/wp-login.php' ). "<br />";
		if( false !== strpos( $REQUEST_URI, '/wp-login.php' ) ) //in wp-login.php hooks ie "add_action( 'admin_footer', & add_action( 'wp_footer'," down work since there is no  <footer>
		{ 
			$OSOLCCC_CommonClass_inst = \OSOLCCC\Hooks\Common::getInstance();
			$OSOLCCC_CommonClass_inst->add_ajaxurl_cdata_to_front();
			$OSOLCCC_CommonClass_inst->add_ccc_onload();
		}
```
5. in MVC/Hooks/Common.php changed `function refreshOSOLMultiCaptchaImage()` so ajax call for captcha image is made only when required
## Stable Tag 0.8

### Made more OOP

### 24-06-2022

*Objective:* 
seperate logic for plugin (further seperate front end and backend) & wp hooks

1. created new folder `Helpers` for plugnin helper classes , name space will be `namespace OSOLCCCF\Helpers`
2. used `spl_autoload_register` to [autoload](https://www.php.net/manual/en/function.spl-autoload-register.php)
```
// auto load Helpers/Frontend.php while calling new \OSOLCCC\Helpers\Frontend() 
if(!function_exists('version_compare') || version_compare(phpversion(), '5.1.0', '<'))die("Minimum version required for 'Customizable Captcha and contact us' plugin is 5.1.0");
spl_autoload_register(function ($class) {
	// project-specific namespace prefix
	$prefix = 'OSOLCCC\\Helpers\\';	
	// base directory for the namespace prefix
	$base_dir = dirname(__FILE__).DIRECTORY_SEPARATOR.'Helpers/';
	//die($class . " " . str_replace($prefix,'',$class));
	// does the class use the namespace prefix?
	$len = strlen($prefix);
	if (strncmp($prefix, $class, $len) !== 0) {
		// no, move to the next registered autoloader
		return;
	}	
	// get the relative class name
	$relative_class = substr($class, $len);
	// replace the namespace prefix with the base directory, replace namespace
	// separators with directory separators in the relative class name, append
	// with .php
	$mappedFile = $base_dir . str_replace('\\', '/', $relative_class) . '.php';
	if (file_exists($mappedFile)) {
		require $mappedFile;
	}
	else
	{
		die('<p style="background:#f00">ERROR!!!!! file : '.$mappedFile. " does not exist to autoload for ".$class ."</p>");
	}
});
```
2. created new `FrontEnd` , `Backend` and `Wphooks` classes in `Helpers` folder
5. declared the following in the beginning.
```
require_once(dirname(__FILE__).DIRECTORY_SEPARATOR.'pluginClasses/FrontEnd.php');
require_once(dirname(__FILE__).DIRECTORY_SEPARATOR.'pluginClasses/BackEnd.php');
require_once(dirname(__FILE__).DIRECTORY_SEPARATOR.'pluginClasses/WPHooks.php');
$OSOL_CCC_HandlerFrontEnd_inst = new \OSOLCCCF\PluginHelpers\FrontEndMain();
$OSOL_CCC_HandlerBackEnd_inst = new \OSOLCCCF\PluginHelpers\BackEndMain();
$OSOL_CCC_HandlerWPHooks_inst = new \OSOLCCCF\PluginHelpers\WPHooks();
```
6. replaced all add_action\('([^']+)',\s*'([^']+)'   with add_action\('\1',[$OSOL_CCC_HandlerFrontEnd_inst,'\2'] 
7. Main file `custCaptchaContact.php` only contains calls to wp hooks




## Stable Tag 0.7

### Made Captcha GDPR Compliant

### 22-06-2022

1. OSOLMultiCaptcha.php replaced with latest [OSOLMultiCaptcha](https://github.com/osolgithub/OSOLMulticaptcha)
2. captcha-customization-options.php
	1. 
	
	<li>
		<label id="jform_params_caseInsensitive-lbl" for="jform_params_caseInsensitive" title="">GDPR Compliant(No Cookie):</label><input type="checkbox" id="jform_params_gdprCompliantNoCookie" name="OSOLMulticaptcha_gdprCompliantNoCookie" value="1" <?php echo get_option('OSOLMulticaptcha_gdprCompliantNoCookie') == 1?"checked=\"checked\"":'';?>/>
	</li>
	
	2. 
	
	<li>
		<?php $OSOLCaptchaEncryptionKey = get_option('OSOLCaptchaEncryptionKey') != ''?get_option('OSOLCaptchaEncryptionKey'):"YourUniqueEncryptionKey";?>
		<label id="jform_params_bgColor-lbl" for="jform_params_bgColor" title="">Encryption Key</label>
		<input name="OSOLCaptchaEncryptionKey" id="jform_params_bgColor" value="<?php echo $OSOLCaptchaEncryptionKey;?>" size="25" type="text" />
    </li>
  
3. displayCaptcha.php: 
	1. added following blocks to prevent setting $_SESSION
	
	```
	$OSOLMulticaptcha_gdprCompliantNoCookie = get_option('OSOLMulticaptcha_gdprCompliantNoCookie');
	if(!$OSOLMulticaptcha_gdprCompliantNoCookie)
	{
	```
	
	and following block to show json object instead of image
	
	```
		if(!$OSOLMulticaptcha_gdprCompliantNoCookie)
		{
			$captchaEncryptionKey = get_option('OSOLCaptchaEncryptionKey');// IMPORTANT ****** YOU MUST SET A CUSTOM VALUE FOR YOUR SITE from admin panel
			$captcha->setCaptchaEncryptionKey($captchaEncryptionKey);
			$returnImgObj = true;
			$captchaImgObj = $captcha->displayCaptcha($returnImgObj);
			ob_start();
			imagepng($captchaImgObj);
			$imageContent = ob_get_contents();
			ob_end_clean(); 
			
			$var2Display = new stdClass();
			$var2Display->captchaEncypted = $captcha->getEncryptedCaptchaString();
			$var2Display->imageContent = base64_encode($imageContent);
			die(json_encode($var2Display));
			
		}
		else
		{
	```	
	

4. custCaptchaContact.php added
	1. added
	```

navtreedata.js
/*
 @licstart  The following is the entire license notice for the JavaScript code in this file.

 The MIT License (MIT)

 Copyright (C) 1997-2020 by Dimitri van Heesch

 Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 and associated documentation files (the "Software"), to deal in the Software without restriction,
 including without limitation the rights to use, copy, modify, merge, publish, distribute,
 sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all copies or
 substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

 @licend  The above is the entire license notice for the JavaScript code in this file
*/
var NAVTREE =
[
  [ "Captcha and Contact Us Forms for Wordpress", "index.html", [
    [ "Change Log", "md_change_log.html", [
      [ "Stable Tag 0.9", "md_change_log.html#autotoc_md1", [
        [ "08-10-2022", "md_change_log.html#autotoc_md2", null ]
      ] ],
      [ "Stable Tag 0.8", "md_change_log.html#autotoc_md3", [
        [ "Made more OOP", "md_change_log.html#autotoc_md4", null ],
        [ "24-06-2022", "md_change_log.html#autotoc_md5", null ]
      ] ],
      [ "Stable Tag 0.7", "md_change_log.html#autotoc_md6", [
        [ "Made Captcha GDPR Compliant", "md_change_log.html#autotoc_md7", null ],
        [ "22-06-2022", "md_change_log.html#autotoc_md8", null ]
      ] ]
    ] ],
    [ "documentationPipeline", "md_documentation_pipeline.html", null ],
    [ "README", "md_doxy_props_readme4_doxygen.html", [
      [ "Author", "md_doxy_props_readme4_doxygen.html#autotoc_md14", null ],
      [ "Contributors", "md_doxy_props_readme4_doxygen.html#autotoc_md15", null ],
      [ "Synopsis", "md_doxy_props_readme4_doxygen.html#autotoc_md16", null ],
      [ "Description", "md_doxy_props_readme4_doxygen.html#autotoc_md17", [
        [ "Plugin to add captcha to core wordpress forms and additional option for contact us page.", "md_doxy_props_readme4_doxygen.html#autotoc_md18", null ],
        [ "Use Case Diagrams", "md_doxy_props_readme4_doxygen.html#autotoc_md19", null ]
      ] ],
      [ "Prerequisites", "md_doxy_props_readme4_doxygen.html#autotoc_md20", null ],
      [ "Installation", "md_doxy_props_readme4_doxygen.html#autotoc_md21", null ],
      [ "Extending / Installing Addons", "md_doxy_props_readme4_doxygen.html#autotoc_md22", null ],
      [ "Contributing", "md_doxy_props_readme4_doxygen.html#autotoc_md23", null ],
      [ "License / Copyright Info", "md_doxy_props_readme4_doxygen.html#autotoc_md24", null ],
      [ "Citation", "md_doxy_props_readme4_doxygen.html#autotoc_md25", null ],
      [ "Contact", "md_doxy_props_readme4_doxygen.html#autotoc_md26", null ]
    ] ],
    [ "Wordpress Plugin Development", "md_wordpress_plugin_development.html", [
      [ "Short Description", "md_wordpress_plugin_development.html#autotoc_md34", null ],
      [ "Detailed Description", "md_wordpress_plugin_development.html#autotoc_md35", [
        [ "Types of extensions possible", "md_wordpress_plugin_development.html#autotoc_md36", null ]
      ] ],
      [ "How system identifies extension", "md_wordpress_plugin_development.html#autotoc_md37", null ],
      [ "Frontend", "md_wordpress_plugin_development.html#autotoc_md38", [
        [ "Detecting activation", "md_wordpress_plugin_development.html#autotoc_md39", null ],
        [ "Interaction with system", "md_wordpress_plugin_development.html#autotoc_md40", null ],
        [ "Display Errors", "md_wordpress_plugin_development.html#autotoc_md41", null ],
        [ "Handling AJAX", "md_wordpress_plugin_development.html#autotoc_md42", null ],
        [ "Using DB", "md_wordpress_plugin_development.html#autotoc_md43", null ],
        [ "Getting User Details", "md_wordpress_plugin_development.html#autotoc_md44", null ],
        [ "Pagination", "md_wordpress_plugin_development.html#autotoc_md45", null ],
        [ "Mailing", "md_wordpress_plugin_development.html#autotoc_md46", null ],
        [ "Javascript builtin options availavle", "md_wordpress_plugin_development.html#autotoc_md47", null ],
        [ "Relative URLs", "md_wordpress_plugin_development.html#autotoc_md48", null ],
        [ "Multilanguage implementation", "md_wordpress_plugin_development.html#autotoc_md49", null ],
        [ "Adding 'adsense'", "md_wordpress_plugin_development.html#autotoc_md50", null ]
      ] ],
      [ "Backend", "md_wordpress_plugin_development.html#autotoc_md51", [
        [ "Adding configuration pages", "md_wordpress_plugin_development.html#autotoc_md52", null ],
        [ "Additional options to interact with plugin", "md_wordpress_plugin_development.html#autotoc_md53", null ],
        [ "Identifying User Levels", "md_wordpress_plugin_development.html#autotoc_md54", null ]
      ] ],
      [ "Crons", "md_wordpress_plugin_development.html#autotoc_md55", [
        [ "Disable from DB", "md_wordpress_plugin_development.html#autotoc_md56", null ]
      ] ],
      [ "Disabling extensions without removing settings", "md_wordpress_plugin_development.html#autotoc_md57", null ],
      [ "Show \"Briefly unavailable for scheduled maintenance. Check back in a minute\"", "md_wordpress_plugin_development.html#autotoc_md58", null ],
      [ "Debugging", "md_wordpress_plugin_development.html#autotoc_md59", null ],
      [ "Troubleshooting", "md_wordpress_plugin_development.html#autotoc_md60", null ],
      [ "Templating", "md_wordpress_plugin_development.html#autotoc_md61", null ]
    ] ],
    [ "Bug List", "bug.html", null ],
    [ "Modules", "modules.html", "modules" ],
    [ "Namespaces", "namespaces.html", [
      [ "Namespace List", "namespaces.html", "namespaces_dup" ]
    ] ],
    [ "Data Structures", "annotated.html", [
      [ "Data Structures", "annotated.html", "annotated_dup" ],
      [ "Data Structure Index", "classes.html", null ],
      [ "Class Hierarchy", "hierarchy.html", "hierarchy" ],
      [ "Data Fields", "functions.html", [
        [ "All", "functions.html", null ],
        [ "Functions", "functions_func.html", null ]
      ] ]
    ] ],
    [ "Files", "files.html", [
      [ "File List", "files.html", "files_dup" ],
      [ "Globals", "globals.html", [
        [ "All", "globals.html", null ],
        [ "Functions", "globals_func.html", null ],
        [ "Variables", "globals_vars.html", null ]
      ] ]
    ] ]
  ] ]
];

var NAVTREEINDEX =
[
"annotated.html"
];

var SYNCONMSG = 'click to disable panel synchronisation';
var SYNCOFFMSG = 'click to enable panel synchronisation';